walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,132 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
+
+  // Returns available times for meeting requesters based on the events and its attendees' schedules.
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> optionalAttendees = new HashSet(request.getOptionalAttendees());
+    Set<String> allAttendees = getAllAttendees(events);
+    List<TimeRange> availableTimeRanges = new ArrayList();
+
+    List<Event> eventsListOfMandatory = createSortedEventsListOfMandatory(events, optionalAttendees);
+    List<Event> eventsListOfMandatoryAndOptional = createSortedEventsListOfMandatoryAndOptional(events);
+
+    if (isRequestDurationTooLong(request)) {
+      return availableTimeRanges;
+    }
+
+    if (hasNoOverlappingAttendees(request.getAttendees(), allAttendees)) {
+      availableTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+      return availableTimeRanges;
+    }
+
+    List<TimeRange> timeRangesList = findTimeRangesOfEvents(eventsListOfMandatoryAndOptional);
+    removeNestedAndOverlappedEvents(timeRangesList);
+    addAvailableTimeRanges(availableTimeRanges, timeRangesList, request);
+
+    if (availableTimeRanges.size() == 0) {
+      List<TimeRange> timeRangesListOnlyMandatory = findTimeRangesOfEvents(eventsListOfMandatory);
```

Comment:
add a comment that you are removing the optional attendees and only looking at the mandatory ones now.